### ISSUE TITLE
Create tickets for refactoring

### ISSUE THREAD

**Issue by kfarr** [3DStreet/3dstreet-editor#40]
Create a few tickets (2 or 3) that outline initial work to do for refactoring the application as verbally suggested in meeting

**Comment by kfarr** [comment]
you can create the tickets directly on this repo or add bullets to this ticket

**Comment by kfarr** [comment]
Update: instead we should see if we can incorporate refactoring from this PR:
https://github.com/aframevr/aframe-inspector/pull/647

**Comment by kfarr** [comment]
In the perfect world we keep `3dstreet-editor` up-to-date with improvements to `aframe-inspector`

An example of this is the PR 647 above from Vincent.

This could also be a way to "shortcut" refactoring to integrate the work that Vincent has done already. If we want to do that there are 2 options
* "rebase" from the updated `aframe-inspector` with merged PR 647 and then re-add our customizations on top
* continue with our fork and "add in" the improvements from PR 647

Why is this important?
* in theory as we and Vincent and other developers add custom features to the aframe-inspector it would be ideal to be able to reuse and share components from each other. An example of this, Vincent is working on a "snap to plane" and "Snap to object" (for example when positioning a car it stays on a plane)

Thoughts from you all?

**Comment by vincentfretin** [comment]
Rebasing and push forcing master may not be the best for you, if you didn't fix a conflict correctly you may lose work. Also if one of the developer didn't reset their master branch on their machine and push force again, it will be a real nightmare. :D 
You should simply do a merge. But please wait that the changes are merged upstream or in a community fork first, in one or two weeks.
You actually don't have much conflicts, but to ease future merge, what you should do is add the `.prettierrc.json` file and convert back to single comma the react files you modified with the old prettier version, the conflicts will be mainly some imports really in the react files, and for the style files, just take your changes only.

**Comment by kfarr** [comment]
Some good questions from @omedvediev :
- What’s your general expectations about refactoring? Do you have some requirements you want us to follow?
- Is it correct, saying you want the project to have an ability to be merged with the original repository?
- Can you describe few next steps of the project’s evolution or evolution direction?

**Comment by kfarr** [comment]
1) General expectations for refactoring
* Honest: it is very hard for me to write checks for dev work that does not improve the user experience. Therefore it's hard for me to imagine being happy with spending 40-80 hours on improvements to the react app, etc. that result in parity for a user.
* On the other hand, as a developer on this project I experience the pain of the "spaghetti code" and older versions of stuff like webpack or whatever build tool this repo is using. 
* So if we mix those together, the most important labor use for refactoring is to improve the quality of life for developers / contributors to this project so that they don't have to spin wheels dealing with build crap and old versions when they're trying to do simple things.

2) Is it correct, saying you want the project to have an ability to be merged with the original repository?
* We have forked this from `aframevr/aframe-inspector`. We have spoken with @dmarcos who maintains that project. He is happy to receive PR that improve the existing "inspector" functionality of the `aframe-inspector` but it is beyond the scope of the inspector to be an editor.
* We have also spoken with @vincentfretin a few times and he is working on a similar project that forks the inspector to create an editor. 
* In the perfect world (which may not be possible, but just hoping) it would be nice to have a unified version with Vincent's and 3DStreet that is also able to send improvements upstream to aframe-inspector. This might be unrealistic but it is a nice dream.

3) Can you describe few next steps of project's evolution or direction?
* The long-term goal is that a non-technical user can use 3DStreet as follows: enter a long/latitude (or street / intersection); have the 3d street view rendered using a 3rd party data source such as https://github.com/a-b-street/osm2lanes or https://github.com/a-b-street/osm2streets; allow the user to change placement of objects such as vehicles, trees, street furniture, materials, buildings etc; allow the user to change the street geometry using "path tools" (add / create / change / modify points of a polygon that represents street features such as curbs, lanes, etc.); allow the user to create different camera views of this as well as basic animation; allow the user to share these experiences in many formats from still images to videos and interactive experiences such as VR.
* The purpose of all of the above is to allow anyone to suggest and improve streetscape improvements with an ultimate goal of creating safer streets for non-automobile use. Implied in the above is the ability a user to create an account and save their streets to modify at a future date and share with others.
* The longer-term goal would be to add "enterprise features" such as SSO, privacy, integration with professional engineering tools, such that for government agencies and consultancies with "willingness to pay" we can charge them money while offering the rest of the application for free for the majority of users.

**Comment by vincentfretin** [comment]
Thanks @kfarr for the summary. For aframe-inspector, @dmarcos just merged https://github.com/aframevr/aframe-inspector/pull/639 that's great, but he still need to review and merge https://github.com/aframevr/aframe-inspector/pull/648 before I rebase and finish https://github.com/aframevr/aframe-inspector/pull/647 After that we will have a solid base with up to date versions for our editor projects.

**Comment by vincentfretin** [comment]
Once the aframe-inspector is in a good state, I was thinking of forking it in the c-frame organization and renaming it to aframe-editor where I'll put behind a flag the drag n drop feature to add new models in the scene. I started implementing it in my project. @kylebakerio was also interested by this feature.

**Comment by omedvediev** [comment]
Hello, @kfarr! 
After some thoughts, according to the listed above in the comments and the goal to fix both UI and code I suggest to make next steps:

- App engine update. At least we should update React and React Dom up to 18v, which will provide us access to React Hooks and Functional Components (cleaner/less code, faster runtime, lots of hooks powers and advantages, such as rerendering control, memoization, etc). In a perfect world, we should update Webpack, but I think it might cause some compatibility issues. So we can update Webpack later.
- UI Kit. Implementing the UI Kit, according to the fact you have an excellent system provided [here](https://www.figma.com/file/W9jid3A0jgssnIKBFEvG38/3DStreet-2022?node-id=1%3A304&t=MpjdWfOzPAaf6mCM-0). It might be a great start to making the app consistent and layout-shift-proof.
- Transfer to SCSS modules (optional). It will give an advantage of using scss modules with SASS  instead of Stylus preprocessor, which in its turn will make it possible to style our components in an isolated manner. It will reduce a lot of css code and styles conflicts.
- Replacing all non-ui-kit components with components from the newly developed UI Kit.

I’m strongly convinced that this refactoring will avoid most of the current layout and styles issues that we have regularly, clean up redundant code, and, finally, update the app to get more modern dev features.

This is what I suggest we do in our first refactoring steps. What are your thoughts on this?

**Comment by vincentfretin** [comment]
[aframe-inspector](https://github.com/aframevr/aframe-inspector) master now is up to date on webpack 5 and react 18 with prettier reformat and all eslint errors fixed. You'll want to merge all the changes in your editor (resolving conflicts and ignoring the css changes most likely) before making any other refactoring.

**Comment by kfarr** [comment]
Thanks @vincentfretin! Great to hear!

@omedvediev Vincent's suggestion to merge changes from aframe-inspector makes sense to me and seems to be compatible with your suggestion for "App engine update...  React and React Dom up to 18v"

Would you be able to merge in the recent aframe-inspector changes while maintaining our specific css and application functionality?

**Comment by vincentfretin** [comment]
I did the merge for you, see #89

**Comment by kfarr** [comment]
Hi @omedvediev , after chatting with Vincent yes it is ok to also do "Transfer to SCSS modules" as part of the refactoring. The next step as discussed is to create an initial list of which components start these changes with

**Comment by omedvediev** [comment]
Hello, @kfarr! 
So, I suggest to start UI-Kit development with next components:

"Atoms":
- Logo (https://www.figma.com/file/W9jid3A0jgssnIKBFEvG38/3DStreet-2022?node-id=700%3A8010&t=bcUje7Jzwr0QqfZ1-4);
- Button (https://www.figma.com/file/W9jid3A0jgssnIKBFEvG38/3DStreet-2022?node-id=698%3A11006&t=bcUje7Jzwr0QqfZ1-4);
- Input (https://www.figma.com/file/W9jid3A0jgssnIKBFEvG38/3DStreet-2022?node-id=700%3A7835&t=bcUje7Jzwr0QqfZ1-4);
- Dropdown (https://www.figma.com/file/W9jid3A0jgssnIKBFEvG38/3DStreet-2022?node-id=700%3A7987&t=bcUje7Jzwr0QqfZ1-4);
- Checkbox (https://www.figma.com/file/W9jid3A0jgssnIKBFEvG38/3DStreet-2022?node-id=700%3A7975&t=bcUje7Jzwr0QqfZ1-4);

And I'll also start SCSS modules implementation simultaneously for those components.

After they are finished, I'll replace all inconsistent components in the App with those from UI-Kit

**Comment by kfarr** [comment]
@omedvediev this is great, thanks for this list. Before starting on this can you help to resolve the conflicts on #88 #87 #70 ?

**Comment by omedvediev** [comment]
Hello, @kfarr , conflicts are resolved, PRs are ready to be merged.

**Comment by kfarr** [comment]
Tracking the 4 new widgets in progress here:
https://github.com/orgs/3DStreet/projects/5/views/1?pane=issue&itemId=19570684

**Comment by kfarr** [comment]
@omedvediev the "atoms" listed above are now complete with the merge of https://github.com/3DStreet/3dstreet-editor/pull/119

what would be some next high prio items for refactoring? cc @ihordubas1

**Comment by kfarr** [comment]
done for now
